zipline-kotlin-plugin/src/main/kotlin/app/cash/zipline/apicheck/ZiplineServicesReader.kt

```diff
@@ -0,0 +1,182 @@
+import org.jetbrains.kotlin.fir.types.FirTypeProjection
+import org.jetbrains.kotlin.fir.types.FirTypeProjectionWithVariance
+import org.jetbrains.kotlin.fir.types.FirTypeRef
+import org.jetbrains.kotlin.fir.types.FirUserTypeRef
+import org.jetbrains.kotlin.types.Variance
+
+fun readZiplineServices(
+  sources: Collection<File>,
+  dependencies: Collection<File>,
+): List<DeclaredZiplineService> {
+  return KotlinFirLoader(sources, dependencies).use { loader ->
+    val output = loader.load("zipline-api-dump")
+    ZiplineServicesReader(output).read()
+  }
+}
+
+private val ziplineFqPackage = FqPackageName("app.cash.zipline")
+private val ziplineServiceClassId = ziplineFqPackage.classId("ZiplineService")
+
+/**
+ * Read the frontend intermediate representation of a program and emit its ZiplineService
+ * interfaces. These are subject to strict API compatibility requirements.
+ */
+internal class ZiplineServicesReader(
+  output: FirResult,
+) {
+  private val platformOutput = output.platformOutput
+  private val session: FirSession = platformOutput.session
+
+  private val ziplineServiceClass: FirClassLikeSymbol<*>? =
+    session.symbolProvider.getClassLikeSymbolByClassId(ziplineServiceClassId)
+
+  fun read(): List<DeclaredZiplineService> {
+    val types = platformOutput.fir
+      .flatMap { it.declarations.findRegularClassesRecursive() }
+      .filter { it.isInterface && it.isZiplineService }
+
+    return types
+      .map { it.asDeclaredZiplineService() }
+      .sortedBy { it.name }
+  }
+
+  private val FirRegularClass.isZiplineService: Boolean
+    get() {
+      val ziplineServiceClssSymbol = ziplineServiceClass as? FirClassSymbol<*> ?: return false
+      return ziplineServiceClssSymbol.isSupertypeOf(symbol, session)
+    }
+
+  private fun FirRegularClass.asDeclaredZiplineService(): DeclaredZiplineService {
+    return DeclaredZiplineService(
+      symbol.classId.asSingleFqName().asString(),
+      bridgedFunctions(this),
+    )
+  }
+
+  private fun bridgedFunctions(type: FirRegularClass): List<DeclaredZiplineFunction> {
+    val result = sortedSetOf<DeclaredZiplineFunction>(
+      { a, b -> a.signature.compareTo(b.signature) }
+    )
+
+    for (supertype in type.getAllSupertypes(session)) {
+      if (!supertype.isInterface) continue // Skip kotlin.Any.
+
+      for (declaration in supertype.declarations) {
+        when (declaration) {
+          is FirFunction -> {
+            if (declaration.isNonInterfaceFunction) continue
+            result += declaration.asDeclaredZiplineFunction()
+          }
+
+          is FirProperty -> {
+            result += declaration.asDeclaredZiplineFunction()
+          }
+
+          else -> Unit
+        }
+      }
+    }
+
+    return result.toList()
+  }
+
+  private val FirFunction.isNonInterfaceFunction: Boolean
+    get() = symbol.name.identifier in NON_INTERFACE_FUNCTION_NAMES
+
+  private fun FirFunction.asDeclaredZiplineFunction(): DeclaredZiplineFunction {
```

Comment:
There’s annoying duplication between this and the compiler plugin. The compiler plugin is IR, this is FIR, and we can’t have nice things.